tensorflow/core/kernels/data/name_utils.h

```diff
@@ -27,39 +27,55 @@ namespace name_utils {
 extern const char kDelimiter[];
 extern const char kDefaultDatasetDebugStringPrefix[];
 
+struct OpNameParams {
+  int op_version = 1;
+};
+
+struct DatasetDebugStringParams {
+  int op_version = 1;
+  string dataset_prefix = "";
+  std::vector<string> args;
+};
+
 // Merge the given args in the format of "(arg1, arg2, ..., argn)".
 //
-// e.g. ArgsToString({"1", "2", "3"}) -> "(1, 2, 3)".
-string ArgsToString(std::initializer_list<StringPiece> args);
+// e.g. ArgsToString({"1", "2", "3"}) -> "(1, 2, 3)"; ArgsToString({}) -> "".
+string ArgsToString(const std::vector<string>& args);
 
 // Returns the dataset op name.
 //
 // e.g. OpName("Map") -> "MapDataset".
 string OpName(const string& dataset_type);
 
+// Returns the dataset op names.
+//
+// e.g. OpName(ConcatenateDatasetOp::kDatasetType, OpNameParams())
+// -> "ConcatenateDataset"
+//
+// OpNameParams params;
+// params.op_version = 2;
+// OpName(ParallelInterleaveDatasetOp::kDatasetType, params)
+// -> "ParallelInterleaveDatasetV2"
+string OpName(const string& dataset_type, const OpNameParams& params);
+
 // Returns a human-readable debug string for this dataset in the format of
 // "FooDatasetOp(arg1, arg2, ...)::Dataset".
 //
-// e.g. DatasetDebugString("Map", "", {}) -> "MapDatasetOp::Dataset";
-// DatasetDebugString("Range", "", {"0", "10", "3"}) ->
-// "RangeDatasetOp(0, 10, 3)::Dataset";
-// DatasetDebugString("Shuffle", "FixedSeed", {"10", "1", "2"}) ->
-// "ShuffleDatasetOp(10, 1, 2)::FixedSeedDataset";
-string DatasetDebugString(const string& dataset_type,
-                          const string& dataset_name_prefix,
-                          std::initializer_list<StringPiece> args);
+// e.g. DatasetDebugString("Map") -> "MapDatasetOp::Dataset";
+string DatasetDebugString(const string& dataset_type);
 
 // Returns a human-readable debug string for this dataset in the format of
 // "FooDatasetOp(arg1, arg2, ...)::Dataset".
 //
-// e.g. DatasetDebugString("Map") -> "MapDatasetOp::Dataset";
-// DatasetDebugString("Range", 0, 10, 3) -> "RangeDatasetOp(0, 10, 3)::Dataset".
-template <typename... Args>
-string DatasetDebugString(const string& dataset_type, const Args&... args) {
-  return DatasetDebugString(
-      dataset_type, kDefaultDatasetDebugStringPrefix,
-      {static_cast<const strings::AlphaNum&>(args).Piece()...});
-}
+// e.g.
+//  DatasetDebugStringParams range_params;
+//  range_params.args.emplace_back(std::to_string(0));
```

Comment:
could we make this work with the following syntax:

```
range_params.args = {0, 2.0, "hello"}
```

If `args` is of the right type (e.g. `AlphaNum`}, then the a constructor should take care of it?

Reply:
@jsimsa  This syntax is good and makes code clean. 

One issue is that `AlphaNum` is not allowed to copy and assign as defined here [TF_DISALLOW_COPY_AND_ASSIGN(AlphaNum)](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/strings/strcat.h#L142), so `args` cannot be initialized in this way `args = {0, 2.0, "hello"}` if it is declared as `std::vector<AlphaNum> args;`.

Reply:
Could we add a function `void AddArgs(const T&... args)` into `Struct DatasetDebugStringParams` so that `params.args` can be initialized by `params.AddArgs(0, 2.0, "hello")`?

Reply:
For example:
```C++
struct DatasetDebugStringParams {
  template <typename... T>
  void AddArgs(T... input_args) {
    args = {static_cast<const strings::AlphaNum&>(input_args).data()...};
  }

  int op_version = 1;
  string dataset_prefix = "";
  std::vector<string> args;
};

name_utils::DatasetDebugStringParams params;
params.AddArgs(0, 2.0, "hello");
```

Reply:
Change the type of the `args` member to `std::initialize_list<strings::AlphaNum>` and then the `{1, 2.0, "hello"}` initialization will work.

Reply:
@jsimsa I did a test using `std::initializer_list<strings::AlphaNum> args = {"1", "2.0", "hello"};`, which reports the following errors:
```
ERROR: /Users/fei/Documents/Github/tensorflow/tensorflow/core/kernels/data/BUILD:90:1: C++ compilation of rule '//tensorflow/core/kernels/data:name_utils_test' failed (Exit 1)
tensorflow/core/kernels/data/name_utils_test.cc:72:52: error: copying array element of type 'tensorflow::strings::AlphaNum' invokes deleted constructor
  std::initializer_list<strings::AlphaNum> args = {1, 2.0, "hello"};
                                                   ^
./tensorflow/core/lib/strings/strcat.h:142:31: note: 'AlphaNum' has been explicitly marked deleted here
  TF_DISALLOW_COPY_AND_ASSIGN(AlphaNum);
                              ^
tensorflow/core/kernels/data/name_utils_test.cc:72:55: error: copying array element of type 'tensorflow::strings::AlphaNum' invokes deleted constructor
  std::initializer_list<strings::AlphaNum> args = {1, 2.0, "hello"};
                                                      ^~~
./tensorflow/core/lib/strings/strcat.h:142:31: note: 'AlphaNum' has been explicitly marked deleted here
  TF_DISALLOW_COPY_AND_ASSIGN(AlphaNum);
                              ^
tensorflow/core/kernels/data/name_utils_test.cc:72:60: error: copying array element of type 'tensorflow::strings::AlphaNum' invokes deleted constructor
  std::initializer_list<strings::AlphaNum> args = {1, 2.0, "hello"};
                                                           ^~~~~~~
./tensorflow/core/lib/strings/strcat.h:142:31: note: 'AlphaNum' has been explicitly marked deleted here
  TF_DISALLOW_COPY_AND_ASSIGN(AlphaNum);
                              ^
3 errors generated.``

Reply:
`std::queue` works in the following way (https://en.cppreference.com/w/cpp/container/queue/emplace):

```
std::queue<strings::AlphaNum> args;
args.emplace(0);
args.emplace(2.0);
args.emplace("hello");
```

but it does not support `braced-init-list` construction (`std::queue<strings::AlphaNum> args = {1, 2.0, "hello"};`).

Reply:
That's unfortunate (it works on my compiler).

If that's the case, let's replace the `args` argument with`dataset_suffix` which is a string and in the places where we use it, we create the string in the dataset op using:

```
params.dataset_suffix = strings::StrCat("(", absl::StrJoin(std::make_tuple(arg_0, ..., arg_n), ", "), ")");
```

Reply:
Actually, I take it back. Let's go with the `AddArgs` route, except let's call the method `set_args`.

Reply:
> Actually, I take it back. Let's go with the `AddArgs` route, except let's call the method `set_args`.

Ok, will do that. We can change it in the future if a better solution comes up.